Have all processes yield self to lifecycle hooks

Th3-M4jor opened the pull request:
Makes it so that Workers will yield self to life cycle hooks and makes some adjustments related to what's public/private in them.

I did not do this for the dispatcher or supervisor as I felt that did not make sense.

closes #513

Th3-M4jor commented on lib/solid_queue/configuration.rb:
```diff
@@ -33,12 +33,15 @@ def instantiate
-      if only_work? then workers
-      else
-        dispatchers + workers + schedulers
+      @configured_processes ||= begin
+        if only_work? then workers
+        else
+          dispatchers + workers + schedulers
+        end
```
Without memoizing this, we would end up with worker ids of `3, 4, 5` instead of `1, 2, 3` in one of the modified tests. Which would be confusing.

rosa commented:
Hey @Th3-M4jor, thanks for this! 

However, I still don't understand the need for an incremental index for each worker 🤔  Could you elaborate more why this is necessary? These indexes would be duplicated across different supervisors (in the case you're running jobs in multiple machines). 

All processes now [have a unique generated name](https://github.com/rails/solid_queue/blob/0a6c3b69872d2ee39072b91e23ceab5327725751/lib/solid_queue/processes/base.rb#L12-L39), [which is used to locate them](https://github.com/rails/solid_queue/blob/0a6c3b69872d2ee39072b91e23ceab5327725751/lib/solid_queue/supervisor.rb#L176), and they carry other identifying information such as the hostname where they're running and their PID. Wouldn't this be more than enough to identify them?

Th3-M4jor commented:
For my particular case having an incremented index for each worker fits better for metrics reporting. It more closely mirrors how Puma identifies each of its forked workers, and for that I inject the k8s pod name to differentiate there.

Though I do see the point that it might not work for everyone's needs, what if I renamed `worker_id` to `worker_index` and added a `worker_name` which returned the unique generated name?

rosa commented:
I think you don't need a `worker_name` 🤔  You can simply access it via `#name`, no? 

About `worker_index`, if this is just for your particular case, I'd rather not add that, it should be something that's useful in general, not just for one particular way of reporting metrics. I think having the worker yielded in the lifecycle hook makes sense, but I don't see the need for having other attributes besides what is already there.

Th3-M4jor commented:
Fair enough, I removed all references to `worker_id` and squashed it down to a single commit. Accessing just the queues should work well enough for my case.

rosa commented:
Cool! I think, for simplicity, we could have all processes do this, not just workers.

Th3-M4jor commented:
Done, and similarly made a few additional things on each process class private if they were only accessed via tests or from within the class itself.

rosa commented on lib/solid_queue/scheduler.rb:
```diff
@@ -24,6 +22,8 @@ def metadata
     end
 
     private
+      attr_reader :recurring_schedule
```
I'd keep this one public, ok to switch to `attr_reader`, but it doesn't need to be private.

rosa commented on lib/solid_queue/worker.rb:
```diff
@@ -24,6 +26,8 @@ def metadata
     end
 
     private
+      attr_reader :pool
```
Same as with `recurring_schedule`, we can keep `pool` public.

rosa commented on README.md:
```diff
@@ -396,8 +398,25 @@ SolidQueue.on_scheduler_stop
+
+SolidQueue.on_worker_stop do |worker|
+  Rails.logger.info "Worker #{worker.name} stopped with queues: #{worker.queues.join(',')}"
+end
```
I know this is just because these are examples, but Solid Queue's log subscriber already logs all this when processes start and stop 🤔  I'm just not sure if it's ok to add examples that would be discouraged because you'd end up with duplicated/redundant logs.

Th3-M4jor replied:
I changed it to instead set a couple values in a somewhat contrived `MyMetricsReporter` class, which hopefully does a better job of showing what someone might want to use it for.

rosa commented:
Nice! Thanks a lot and sorry for the delay!
